packages/core/contracts/optimistic-assertor/implementation/examples/DataAsserter.sol

```diff
@@ -0,0 +1,104 @@
+import "../../../common/implementation/AncillaryData.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+// This contract allows assertions on any form of data to be made using the UMA Optimistic Assertor and stores the
+// proposed value so that it may be retrieved on chain. It only enables one assertion per pair of dataId and asserter
+// address. The dataId is intended to be an arbitrary value that uniquely identifies a specific piece of information in
+// the consuming contract and is replaceable. Similarly, any data structure can be used to replace the asserted data.
+contract DataAsserter {
+    using SafeERC20 for IERC20;
+    IERC20 public immutable defaultCurrency;
+    OptimisticAssertorInterface public immutable oa;
+    uint256 public constant assertionLiveness = 7200;
+    bytes32 public immutable defaultIdentifier;
+
+    struct DataAssertion {
+        bytes32 dataId; // The dataId that was asserted.
+        bytes32 data; // This could be an arbitrary data type.
+        address asserter; // The address that made the assertion.
+        bool resolved; // Whether the assertion has been resolved.
+    }
+
+    mapping(bytes32 => bytes32) public oaIdsToInternalIds;
+
+    mapping(bytes32 => DataAssertion) public assertionsData;
+
+    event DataAsserted(bytes32 indexed dataId, bytes32 data, address indexed asserter, bytes32 oaAssertionId);
+
+    event DataAssertionResolved(bytes32 indexed dataId, bytes32 data, address indexed asserter, bytes32 oaAssertionId);
+
+    constructor(address _defaultCurrency, address _optimisticAssertor) {
+        defaultCurrency = IERC20(_defaultCurrency);
+        oa = OptimisticAssertorInterface(_optimisticAssertor);
+        defaultIdentifier = oa.defaultIdentifier();
+    }
+
+    // For a given dataId and asserter, returns a boolean indicating whether the data is accessible and the data itself.
+    function getData(bytes32 dataId, address asserter) public view returns (bool, bytes32) {
+        bytes32 dataAssertionId = getAssertionId(dataId, asserter);
+        if (!assertionsData[dataAssertionId].resolved) return (false, 0);
+        return (true, assertionsData[dataAssertionId].data);
+    }
+
+    // Asserts data for a specific dataId on behalf of an asserter address.
+    function assertDataFor(
+        bytes32 dataId,
+        bytes32 data,
+        address asserter
+    ) public returns (bytes32 oaAssertionId) {
+        bytes32 dataAssertionId = getAssertionId(dataId, asserter);
+        asserter = asserter == address(0) ? msg.sender : asserter;
+        require(assertionsData[dataAssertionId].asserter == address(0), "Data already asserted");
+        uint256 bond = oa.getMinimumBond(address(defaultCurrency));
+        defaultCurrency.safeTransferFrom(msg.sender, address(this), bond);
+        defaultCurrency.safeApprove(address(oa), bond);
+        oaAssertionId = oa.assertTruthFor(
+            abi.encodePacked(
+                "Data asserted for dataId: 0x",
```

Comment:
Maybe for completeness its worth also adding asserted `data` to the claim.

Reply:
Actually `data` could be any data structure, for example one of the users is going to use `uint[4]`. That's why I didn't see the point of adding this information to the claim. Do you think it can be left out or would you still put it in?

Reply:
I assume the default `ASSERT_TRUTH` identifier would only include instructions for voters to first get `assertionId` from ancillary data and then finding the relevant `AssertionMade` event to read emitted `claim`. It would be too much to ask from voters to also inspect the calling contract's source code to infer where to look for the asserted `data`. This would only work with non-default price identifier that could add additional instructions to read `assertionsData[oaIdsToInternalIds[assertionId]].data` from the requesting contract.

Also, there is no way for voters to interpret the underlying value unless the data type is somehow encoded in the `bytes32 data`, but that would require custom price identifier to provide instructions on how to decode `data` to its underlying type and real asserted value(s) that were encoded as `bytes32`. So for this example implementation I would consider sufficient just to add `bytes32 data` to the claim with further comments in the documentation on how this could be used for other underlying data types in real life.

Reply:
I had not considered the restriction that UMIP would put on the ASSERT TRUTH to limit the effort that the system's players would have to undertake to verify the data. Both the UMIP limitation and putting everything necessary in the claim to be adequate to reply make a lot of sense now.

I'm going to do as you suggest! Thank you